### DISCUSSION TITLE
chainId of length 3 characters does not work

### DISCUSSION THREAD

**Discussion by Sotatek-ToanTran** [CoolBitX-Technology/coolwallet-signing-tools#28] [Q&A]
# chainId of length 3 characters does not work

- **Description**
we are suspecting the case where chainId is 3 characters length, it will be compile error.
> **_When we test with 2 character chainId everything works back to normal. When we debug and find out more because the chainId is wrong, the parse address is also wrong._**

- **Expected Behavior**
Check the logic of handling the chainId and pass this error.

- **Screenshots or appendix**
![image](https://user-images.githubusercontent.com/31186688/149301557-70124fbf-fd53-4ea1-8793-771ffe59c7aa.png)

![](https://user-images.githubusercontent.com/31679583/149267642-25471a3e-5acd-416c-b413-0e8392db9159.png)

**Comment by Sotatek-ToanTran** [comment]
We try compare rawTx/input & decrypted rawTx, input 137 but decrypted rawTx is 9
![image (1)](https://user-images.githubusercontent.com/31679583/149303850-53659bbc-d671-40cf-942a-eeab924598fd.png)

**Reply by hank121314** [reply]
Could you provide the detail of `transaction`, which is generated by `scriptUtils.getTransferArgument` in `coin-eth`?

**Reply by Sotatek-KhanhLe** [reply]
argument generated by 
![image](https://user-images.githubusercontent.com/88299386/149311522-eb1e2fc5-87aa-4d0e-afaf-475b187b7041.png)
 is 15328000002c8000003c80000000000000000000000081bb32e4A7e4d0500d11A52F3a5F60c9A6Ef126C000000038d7ea4c680000000000000020c8558000000000000000000520c000000000000000000000089
chainId padStart from 8 since I set argChainId to 4 
![image](https://user-images.githubusercontent.com/88299386/149311744-5c2f3750-a654-488c-a483-adda99c14ab0.png)

**Reply by hank121314** [reply]
Could you also provide the `script`? Thanks!

**Reply by Sotatek-KhanhLe** [reply]
On the code snipet above we have already provided the script, 03000601C707000000003CCC0710C0C2ACD70032FFF8C2ACD7001EFFF6C2ACD70028FFF6CC071094CAA02700C2A2D700FFF6CC071080C2ACD7003AFFFCCC07C0028080C3709710DC07C0054D41544943CC0FC0023078BAA02F6C0E04DDF09700DAA2D7C0FFF612D207CC05065052455353425554546F4E

**Comment by hank121314** [comment]
The rawTx above seems to be a `SignedRawTransaction`, which is not generated by our coolwallet.
Seems that there are some mistakes when trying to encode the rawTx and signature to `SignedRawTransaction`.

I use the given `script` and `argument`.
And I get the rawTx with `apdu.getSignedHex`.
Here is the output:
`ec8085020c85580082520c9481bb32e4a7e4d0500d11a52f3a5f60c9a6ef126c87038d7ea4c680008081898080`.
And I also verify the signature, which is generated by `apdu.tx.executeScript` is valid.

If the output of `apdu.getSignedHex` is not the same as above or the signature is not valid. Please provide a reproducible example that triggers the bug.

Thanks!

**Reply by Sotatek-KhanhLe** [reply]
The potential issue might be in this code snipet, when the transaction.chainId is greater than 110 vValue would be greater than 255 then Buffer.from([vValue]) will return wrong "v" value that it should be
```
    const vValue = v + transaction.chainId * 2 + 8;
    const signedTransaction = rawTx.slice(0, 6);
    signedTransaction.push(Buffer.from([vValue]), Buffer.from(r, 'hex'), Buffer.from(s, 'hex'));
    const serializedTx = rlp.encode(signedTransaction);
 ```

**Reply by hank121314** [reply]
It is recommended to read [Buffer](https://nodejs.org/api/buffer.html#bufwriteint16bevalue-offset) document in node.js to resolve this  overflow issue.

This is not our responsibility to resolve this kind of issue. The reported issue should be only coolwallet related.
